File: html2md/main.py
#!/usr/bin/env python3
"""html2md — CLI entry point for HTML → Markdown conversion."""

import argparse
import sys

from lib.md_converter import convert, get_converter


def main():
    ap = argparse.ArgumentParser(
        prog="html2md",
        description="Convert HTML to Markdown.",
    )
    ap.add_argument("--file", help="Local HTML file path")
    ap.add_argument("--html", help="Inline HTML string")
    ap.add_argument("--output", help="Write Markdown to file (default: stdout)")
    ap.add_argument(
        "--backend", default="markdownify",
        choices=["markdownify", "html2text"],
        help="Converter backend (default: markdownify)",
    )
    ap.add_argument("--strip-images", action="store_true", help="Omit <img> tags")
    ap.add_argument("--strip-links", action="store_true", help="Omit <a> tags")
    ap.add_argument("--wrap", type=int, default=0, help="Line wrap width (0=no wrap)")

    args = ap.parse_args()

    # --- resolve input ---
    if args.file:
        try:
            html = open(args.file, encoding="utf-8").read()
        except FileNotFoundError:
            print(f"Error: file not found: {args.file}", file=sys.stderr)
            sys.exit(1)
    elif args.html:
        html = args.html
    else:
        if sys.stdin.isatty():
            ap.print_help(sys.stderr)
            sys.exit(1)
        html = sys.stdin.read()
        if not html.strip():
            print("Error: no input provided. Use --file, --html, or pipe via stdin.", file=sys.stderr)
            sys.exit(1)

    # --- convert ---
    try:
        md = convert(html, backend=args.backend)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)

    # --- wrap ---
    if args.wrap > 0:
        md = _wrap(md, args.wrap)

    # --- output ---
    if args.output:
        open(args.output, "w", encoding="utf-8").write(md + "\n")
        print(f"→ {args.output}", file=sys.stderr)
    else:
        sys.stdout.write(md + "\n")


def _wrap(text: str, width: int) -> str:
    """Simple line-wrap preserving blank lines and Markdown structure."""
    import textwrap
    lines = text.splitlines()
    wrapped = []
    for line in lines:
        if not line.strip():
            wrapped.append("")
        elif line.startswith("#") or line.startswith("-") or line.startswith("|"):
            wrapped.append(line)  # skip headings, lists, tables
        else:
            wrapped.extend(textwrap.wrap(line, width))
    return "\n".join(wrapped)


if __name__ == "__main__":
    main()

File: html2md/src/main.rs
use std::io::{self, Read};
use std::process;

use clap::{CommandFactory, Parser};

use html2md::engine;

#[derive(Parser, Debug)]
#[command(name = "html2md", about = "Convert HTML to Markdown")]
struct Args {
    /// Local HTML file path
    #[arg(long)]
    file: Option<String>,

    /// Inline HTML string
    #[arg(long)]
    html: Option<String>,

    /// Write Markdown to file (default: stdout)
    #[arg(long)]
    output: Option<String>,

    /// Converter backend
    #[arg(long, default_value = "comrak")]
    backend: String,

    /// Omit <img> tags
    #[arg(long)]
    strip_images: bool,

    /// Omit <a> tags
    #[arg(long)]
    strip_links: bool,

    /// Line wrap width (0 = no wrap)
    #[arg(long, default_value = "0")]
    wrap: usize,
}

fn main() {
    let args = Args::parse();

    // --- resolve input ---
    let html = if let Some(ref path) = args.file {
        std::fs::read_to_string(path).unwrap_or_else(|e| {
            eprintln!("Error: file not found: {path}: {e}");
            process::exit(1);
        })
    } else if let Some(ref s) = args.html {
        s.clone()
    } else if atty::is(atty::Stream::Stdin) {
        let _ = Args::command().print_help();
        eprintln!();
        process::exit(1);
    } else {
        let mut buf = String::new();
        io::stdin().read_to_string(&mut buf).unwrap_or_else(|e| {
            eprintln!("Error: {e}");
            process::exit(1);
        });
        if buf.trim().is_empty() {
            eprintln!("Error: no input provided. Use --file, --html, or pipe via stdin.");
            process::exit(1);
        }
        buf
    };

    // --- convert ---
    let md = match engine::convert(&html, &args.backend) {
        Ok(m) => m,
        Err(e) => {
            eprintln!("Error: {e}");
            process::exit(1);
        }
    };

    // --- post-filter ---
    let md = apply_filters(md, &args);

    // --- wrap ---
    let md = if args.wrap > 0 {
        wrap_lines(&md, args.wrap)
    } else {
        md
    };

    // --- output ---
    if let Some(out_path) = args.output {
        std::fs::write(&out_path, &md).unwrap_or_else(|e| {
            eprintln!("Error writing output: {e}");
            process::exit(1);
        });
        eprintln!("→ {out_path}");
    } else {
        print!("{md}");
    }
}

fn apply_filters(md: String, args: &Args) -> String {
    let mut result = md;
    if args.strip_images {
        // Remove ![alt](src) patterns
        result = regex::Regex::new(r"\!\[.*?\]\(.*?\)")
            .unwrap()
            .replace_all(&result, "")
            .to_string();
    }
    if args.strip_links {
        // Remove [text](url) → text
        result = regex::Regex::new(r"\[(.*?)\]\(.*?\)")
            .unwrap()
            .replace_all(&result, "$1")
            .to_string();
    }
    result
}

fn wrap_lines(md: &str, width: usize) -> String {
    let mut wrapped = Vec::new();
    for line in md.lines() {
        if line.trim().is_empty()
            || line.starts_with('#')
            || line.starts_with('-')
            || line.starts_with('|')
            || line.starts_with('>')
        {
            wrapped.push(line.to_string());
        } else if line.len() > width {
            wrapped.extend(word_wrap(line, width));
        } else {
            wrapped.push(line.to_string());
        }
    }
    wrapped.join("\n")
}

fn word_wrap(line: &str, width: usize) -> Vec<String> {
    let mut lines = Vec::new();
    for chunk in line.split_whitespace() {
        if lines.is_empty() {
            lines.push(chunk.to_string());
            continue;
        }
        let last = lines.last_mut().unwrap();
        if last.len() + 1 + chunk.len() <= width {
            last.push(' ');
            last.push_str(chunk);
        } else {
            lines.push(chunk.to_string());
        }
    }
    lines
}
